File: 38. Count and Say.cpp
class Solution {
public:
    string countAndSay(int n) {
        string s = "1";
        for (int i = 2; i <= n; i++) {
            char target;
            int cnt = 0, pos = 0;
            string tmp = "";
            while (pos < s.size()) {
                if (cnt == 0) {
                    target = s[pos];
                    cnt = 1;
                } else if (s[pos] == target) {
                    cnt++;
                } else {
                    tmp += to_string(cnt) + target;
                    cnt = 1;
                    target = s[pos];
                }
                pos++;
            }
            tmp += to_string(cnt) + target;
            s.swap(tmp);
        }
        return s;
    }
};

File: java/38. Count and Say.java
//
class Solution {
    public String countAndSay(int n) {
        StringBuilder cur = new StringBuilder();
        StringBuilder next;
        cur.append("1");
        for (int i = 1; i < n; ++i) {
            next = new StringBuilder();
            int j = 0;
            while (j < cur.length()) {
                int k = j + 1;
                while (k < cur.length() && cur.charAt(k) == cur.charAt(k - 1)) {
                    ++k;
                }
                next.append(k - j);
                next.append(cur.charAt(j));
                j = k;
            }
            cur = next;
        }
        return cur.toString();
    }
}
